src/accessibility/browser.ts

```diff
@@ -0,0 +1,77 @@
+import {
+  getRole,
+  computeAccessibleName,
+  computeAccessibleDescription,
+} from 'dom-accessibility-api';
+import type { AccessibilityTreeOptions } from '.';
+
+const indent = (text: string, indenter = '  ') =>
+  indenter + text.split('\n').join(`\n${indenter}`);
+
+const enum AccessibilityState {
+  /** This element is accessible and its descendents could be accessible */
+  Accessible,
+  /** This element is innaccessible, but its descendents could be accessible */
+  InaccessibleSelf,
+  /** Both this element and its descendents are inaccessible */
+  InaccessibleSelfAndDescendents,
+}
+
+// TODO in PR: what about role="presentation" or role="none"? Should they be excluded?
```

Comment:
Flagging this for review.

My brain has only parsed this like 20% so I don't really understand it yet. The current implementation in this PR is definitely wrong. I will either make it match the spec in this PR or in a follow-up.

If somebody understands this well (w.r.t. how to treat child elements) feel free to explain it to me!

https://www.w3.org/TR/wai-aria-1.2/#tree_exclusion
> The following elements are not exposed via the accessibility API and user agents MUST NOT include them in the accessibility tree:
> ...
> - Elements with none or presentation as the first role in the role attribute. However, their exclusion is conditional. In addition, the element's descendants and text content are generally included. These exceptions and conditions are documented in the presentation (role) section.


https://www.w3.org/TR/wai-aria-1.2/#presentation

> The intended use is when an element is used to change the look of the page but does not have all the functional, interactive, or structural relevance implied by the element type, or may be used to provide for an accessible fallback in older browsers that do not support WAI-ARIA.
> For any element with a role of presentation and which is not focusable, the user agent MUST NOT expose the implicit native semantics of the element (the role and its states and properties) to accessibility APIs. However, the user agent MUST expose content and descendant elements that do not have an explicit or inherited role of presentation. Thus, the presentation role causes a given element to be treated as having no role or to be removed from the accessibility tree, but does not cause the content contained within the element to be removed from the accessibility tree.

Reply:
Oh yeah, weird. MDN makes it even more complicated:

> The presentation role is used to remove semantic meaning from an element and any of its related child elements. For example, a table used for layout purposes could have the presentation role applied to the table element to remove any semantic meaning from the table element and any of its table related children elements, such as table headers and table data elements. Non-table related elements should retain their semantic meaning, however.

Happy to chat more about this, though your quotes made me realize I've been misunderstanding `role="presentation"` (and made me grateful for just using `aria-hidden` instead 😅 )

Reply:
You may want to flag this in our internal #a11y channel or the A11y slack to get some more expertise

Reply:
FWIW I'm planning to research/ask about this and address it in a follow-up PR

Reply:
A question I have is what is the goal here?

For example, is the goal of the accessibility tree test to validate I purposefully added a `presentation` `role` to an HTML element? If yes, then it seems it'd be helpful to have this information included in the snapshot. 🤔 

Did I make sense?

Reply:
@gerardo-rodriguez Good point. Here is how I think about it:

If role="presentation" worked the same way as role="none" (which I don't know if they are but just for the sake of explanation), I would consider which one I was using to be an implementation detail, so they snapshots should be the same regardless of which I'm using. So this makes me reluctant to display a presentation role for `role="presentation"` like we do for other roles:

```
region
  presentation
    heading "hello"
    button "click"
```

It's unclear to me exactly what role="presentation" does to descendent elements. From the MDN page Paul quoted, it seems like descendents that are related to the element with role="presentation" get their roles removed? And maybe other things happen too? Ideally the snapshot would encompass all the changes to the accessibility tree that are described in the spec

Reply:
> FWIW I'm planning to research/ask about this and address it in a follow-up PR

I am wondering if we should have a quick Slack call to better understand and/or resolve this with @cloudfour/dev? Or, @calebeby, do you feel your original request/questions have been answered?

Reply:
I'd be happy to chat. I'm also curious how the presentation role affects its children, and how we should display that info

Reply:
Yeah I think a chat would be good. I could do tomorrow afternoon anytime?

Reply:
Same :+1: